src/display.rs

```diff
@@ -55,12 +55,35 @@ impl I2CDisplay {
     /// `device_id` (normally 1 or 2).
     pub fn new(device_id: u8) -> I2CDisplay {
         let device_path = format!("/dev/i2c-{}", device_id);
-        let d = LinuxI2CDevice::new(device_path, ADDRESS).unwrap();
+        let mut d = LinuxI2CDevice::new(device_path, ADDRESS).unwrap();
         let mut display = I2CDisplay {
             device: d,
             frame: 0,
         };
-        display.register(CONFIG_BANK, MODE_REGISTER, PICTURE_MODE);
+
+        // Display initialization
+        display.reset();
+        display.show(&[[0; 7]; 17]);
+
+        display.device.smbus_write_byte_data(BANK_ADDRESS, CONFIG_BANK);
+        // Switch to Picture Mode
+        display.device.smbus_write_byte_data(MODE_REGISTER, PICTURE_MODE);
+        // Disable audio sync
+        display.device.smbus_write_byte_data(AUDIOSYNC_REGISTER, 0);
+
+        display.device.smbus_write_byte_data(BANK_ADDRESS, 1);
+        display.device.smbus_write_block_data(
+            0,
+            &[255; 17]
+        );
```

Comment:
Turns out I either found a bug in `smbus_write_block_data`, or it just doesn't do what I think it does. Rewriting this to use `smbus_write_byte_data` as below works.

```Rust
for i in 0..17 {
    display.device.smbus_write_byte_data(ENABLE_OFFSET + i, 255);
}
```

Though, we could use `127`, instead of `255` since the final LED of each "row" in each matrix isn't actually hooked up to anything.

Reply:
...another possibility is that I've got some dodgy connections on my board from some sloppy soldering, and bulk I2C transfers aren't reliable as a result. I should go back over the solder joints to make sure this isn't the case.

Reply:
FWIW I remember trying similar patterns (extracted from the datasheet and from the python library) and ended up with the same issue of the missing column. I start suspecting a bug in the I2C library too, but that would be odd as transferring the display buffer in bulk seems to work fine.

Reply:
The only current usage of `smbus_write_block_data` is in `frame`, `sleep`, and `new` (via `register`), and they're all only writing a single byte.

If I update the `show` method to try to write 32 bytes at a time, instead of writing a single byte at a time, I get some very interesting output on Matrix A (everything is shifted down a pixel, and every other pixel on the top row is on), but Matrix B is fine.

```diff
@@ -122,7 +136,7 @@ impl Display for I2CDisplay {
         // TODO(tzn): Double buffering.
         // let new_frame = (self.frame + 1) % 2;
         let new_frame = 1;
-        self.bank(new_frame);
+        let mut buff = [0; 136];
         for y in 0..DISPLAY_HEIGHT {
             for x in 0..DISPLAY_WIDTH {
                 let offset = if x >= 8 {
@@ -134,10 +148,15 @@ impl Display for I2CDisplay {
                     Some(column) => column[y as usize],
                     None => 0,
                 };
-                self.device
-                    .smbus_write_byte_data(COLOR_OFFSET + offset as u8, value);
+                buff[offset] = value;
             }
         }
+        self.bank(new_frame);
+        let mut offset = 0;
+        for chunk in buff.chunks(32) {
+            self.device.smbus_write_block_data(COLOR_OFFSET + offset, chunk);
+            offset += 32;
+        }
```

Changing this version to write a single byte at a time using `smbus_write_block_data` results in an even more corrupted display for me (every pixel is at least partially on).

Reply:
Yeah, this is definitely some wonkiness with `smbus_write_block_data`. This works just fine:

```diff
@@ -122,7 +136,7 @@ impl Display for I2CDisplay {
         // TODO(tzn): Double buffering.
         // let new_frame = (self.frame + 1) % 2;
         let new_frame = 1;
-        self.bank(new_frame);
+        let mut buff = [0; 136];
         for y in 0..DISPLAY_HEIGHT {
             for x in 0..DISPLAY_WIDTH {
                 let offset = if x >= 8 {
@@ -134,10 +148,15 @@ impl Display for I2CDisplay {
                     Some(column) => column[y as usize],
                     None => 0,
                 };
-                self.device
-                    .smbus_write_byte_data(COLOR_OFFSET + offset as u8, value);
+                buff[offset] = value;
             }
         }
+        self.bank(new_frame);
+        let mut offset = 0;
+        for byte in buff.iter() {
+            self.device.smbus_write_byte_data(COLOR_OFFSET + offset, *byte);
+            offset += 1;
+        }
         self.frame(new_frame);
         self.frame = new_frame;
     }
```